New package: pinephone-base-0.1

jjsullivan5196 opened the pull request:
Refer to void-mklive https://github.com/void-linux/void-mklive/issues/119#issuecomment-669163642 for more information

Tested features:

- [x] USB-C connectivity (DisplayPort alt mode, OTG, Power Delivery, Hubs, etc)
- [x] Audio Mixing
  - [x] HDMI Audio
  - [x] Headphone/rear speaker
  - [ ] In-call audio routing (modem codec does work)
- [x] Wifi b/g/n
- [x] Bluetooth connection
- [x] Power management/suspend
- [x] Camera (Activates, but couldn't get MPV to open it.)
- [ ] Modem
  - [x] Power management [(megi's modem driver)](https://xnux.eu/devices/feature/modem-pp.html#toc-modem-power-driver)
  - [ ] 4G data
  - [ ] Calls/sms

sgn commented on srcpkgs/pinephone-kernel/template:
```diff
@@ -0,0 +1,269 @@
+
+	mkdir -p ${hdrdest}/arch/${arch}/kernel
+	cp arch/${arch}/Makefile ${hdrdest}/arch/${arch}
+	if [ "$subarch" = "i386" ]; then
+		mkdir -p ${hdrdest}/arch/x86
+		cp arch/x86/Makefile_32.cpu ${hdrdest}/arch/x86
+	fi
+	if [ "$arch" = "x86" ]; then
+		mkdir -p ${hdrdest}/arch/x86/kernel
+		cp arch/x86/kernel/asm-offsets.s ${hdrdest}/arch/x86/kernel
```
This part needs some clean up

jjsullivan5196 replied:
Fixed this, only kept the branch for aarch64 by itself

ericonr commented on srcpkgs/pinephone-base/template:
```diff
@@ -0,0 +1,11 @@
+archs="aarch64*"
+build_style=meta
+depends="pinephone-kernel linux-firmware pinephone-uboot rtl8723bt-firmware"
+short_desc="Void Linux Pinebook platform package"
```
Pinephone

jjsullivan5196 replied:
fixed this

jjsullivan5196 commented:
> * I failed to see how `or1k` things related to pinephone, I only skimmed over, did I miss anything?

@sgn this is necessary to build power-management firmware for the device, as it uses a custom CPU core running an OpenRISC ISA. More details here https://linux-sunxi.org/AR100 and here https://github.com/crust-firmware/crust

jjsullivan5196 commented:
I got a few more hours at work, I'll resolve these problems by tonight

jjsullivan5196 commented:
@sgn OK, I removed the ppc patch and split the package commits. I left the cross-or1k toolchain as one commit, as  both packages perform one task together for this pull. If this isn't favored, let me know and I'll split out the base pacakge and gcc/binutils.

jjsullivan5196 commented:
I force pushed a couple of minor changes

- Bluetooth firmware is now correctly installed
- Bootargs are added for the state of the volume keys, which can be used for extra boot modes if desired

jjsullivan5196 commented:
Alright, I'm going to take this out of draft as I'm pretty confident most stuff is working. Hardware features tested:

- USB-C connectivity (DisplayPort alt mode, OTG, Power Delivery, Hubs, etc)
- Wifi b/g/n
- Bluetooth connection
- Power management (Still need to look at getting deep sleep working, but I think everything needed for it is there. Probably just needs configuration.)
- Camera (Activates, but couldn't get MPV to open it.)

Only thing I skipped for now is the modem, but it works with similar patchsets in other distros (Mobian, Manjaro, postmarketOS) It'll be easier to test once myself or someone else can package/configure some telephony software.

The travis check is still failing on a couple of issues

- musl can't build uboot for lack of a [stddef.h](https://travis-ci.org/github/void-linux/void-packages/jobs/716255444#L17658) in the include path, or some related reason. I've seen patches for this in the kernel packages, but not sure how to amend it here.
- can't build [or1k gcc on i686](https://travis-ci.org/github/void-linux/void-packages/jobs/716255437#L8594), some generated assembler has an out of bounds issue.

jjsullivan5196 commented:
I think I figured out the uboot problem, I used pinebookpro's uboot as a template and forgot to copy a small detail over. I think travis should be OK with this one.

The only "problem" then is the fact that i686 can't build the or1k cross toolchain for some reason, but I'm not sure why. The build output isn't that helpful.

jjsullivan5196 commented:
So it seems that the or1k cross compiler wasn't meant to run on i686 for some reason, marking it as broken has allowed the travis check to pass.

sgn commented on srcpkgs/pinephone-uboot/files/kernel.d/uboot:
```diff
@@ -0,0 +1,53 @@
+#!/bin/sh
+
+kver=${2}
+uuid=$(cat /etc/fstab | awk '$2 == "/" { print $1 }' | sed 's/^UUID=//')
+dev=$(blkid -U "${uuid}")
+partuuid=$(blkid -o value -s PARTUUID "${dev}")
+bootpart=$(df -P /boot | tail -1 | awk '{ print $6 }')
```
Sorry, this part is a bit complicated.
And it doesn't account for multiple layer fs, (`lvm`, `luks`), right?

*I think* it's better to ask user specify `bootargs` directly in a configuration file.
It seems like other bootloader don't try to guess this value. :-)

I don't have PinePhone, so I could mis-understand everything I said, though :)

sgn replied:
Even if we agree of this logic, is it better to use:
```
df -P / | awk 'NR == 2 {print $1}'
```
instead of relying on forcing on `UUID=` on `/etc/fstab`?

jjsullivan5196 replied:
I was thinking of having something like a `/etc/default/cmdline` for specifying boot args, but I didn't want to stray too far from the pinebookpro template. I agree that it would be better to make this more end-user friendly, so I'm willing to try something else.

As for the complicated part, I 100% agree. I tried something a little less hacky by using blkid's `eval` output mode, which just spitss out conf style variable assignments eg `PARTUUID=deadbeef` that can be sourced by sh.

I personally think the best approach for flexibility reasons would be to source cmdline arguments from `/etc/default/cmdline` and have this configure hook fill it into the boot script. It sits inline with what `grub-mkconfig` does with `/etc/default/grub`, and `void-mklive` could just append `root=...` when creating base images.

jjsullivan5196 replied:
The reconfigure hook now reads settings from `/etc/default/pinephone-uboot-config` to determine the cmdline and toggle cmdline parameters for the volume rocker state.

jjsullivan5196 commented:
Force-pushed to resolve uboot config issue brought up by @sgn. I left details in the review comments.

sgn commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,114 @@
+	export BL31=$wrksrc/$_armtf/build/sun50i_a64/debug/bl31.bin
+	export SCP=$wrksrc/$_crust/build/scp/scp.bin
+
+	if [ "$XBPS_TARGET_LIBC" = "musl" ]; then
+		touch $_uboot/include/stddef.h
+	fi
+
```
Waiting for https://github.com/void-linux/void-packages/pull/24307

jjsullivan5196 replied:
@sgn should I remove this and rebase when #24307 is merged?

ericonr commented:
> So it seems that the or1k cross compiler wasn't meant to run on i686 for some reason (needs 64bit wide integers supported by the host), marking it as broken has allowed the travis check to pass.

All platforms should support 64-bit integers, unless they did some really bad trick. Anyway, mark it broken for platforms with wordsize 32 instead of just i686.

sgn commented:
> > +	if [ "$XBPS_TARGET_LIBC" = "musl" ]; then
> +		touch $_uboot/include/stddef.h
> +	fi
> +
> 
> @sgn should I remove this and rebase when #24307 is merged?

I would say yes, since this PR couldn't build with Travis anyway.
Hence, the final correctness wins the current buildable status.

-- 
Danh

jjsullivan5196 commented:
Removed the musl workaround and all 32bit archs are marked as broken for cross-or1k.

Also @sgn the build was successful a few times, it just sometimes hits the time limit for travis. Hit and miss unfortunately.

sgn commented on srcpkgs/pinephone-kernel/patches/smsc95xx-mac.patch:
```diff
@@ -0,0 +1,95 @@
+From e36d607d56dc5c0cbf2cb600e7686b559ea77b44 Mon Sep 17 00:00:00 2001
+From: popcornmix <popcornmix@gmail.com>
```
Does pinephone have `smsc95xx-mac` device?
Otherwise, I don't find anything obviously wrong, and #24307 is merged, I'll merge this after building.

jjsullivan5196 replied:
I kept it from the 5.7 patches since it's for a USB networking device. Someone could potentially use it thru a USB hub, unless I misunderstood what the `/net/usb` category is for. If it is a misunderstanding I'll delete it.

jjsullivan5196 replied:
I'll keep this in for now as it does seem that these are for USB attached devices, but I can still remove it if it's the right thing to do. https://github.com/torvalds/linux/blob/master/drivers/net/usb/Kconfig

sgn replied:
From what I understand, it's a problem when booting over NFS in the past,
the patch was rejected upstream [1], since then, firmware people had fixed the firmware [2]


1: https://netdev.vger.kernel.narkive.com/Fi2roLel/patch-smsc95xx-add-module-parameter-to-override-mac-address
2: https://github.com/raspberrypi/firmware/issues/613#issuecomment-282681283

jjsullivan5196 replied:
Makes sense, I'll drop it now

jjsullivan5196 commented:
Alright, the rebase seems to have worked. Travis is still kicking out because the build takes too long (varies between 45-50minutes on travis), but I built aarch64-musl on my PC and everything's working.

jjsullivan5196 commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,110 @@
+# branch of arm-trusted-firmware.
+
+_commit_uboot=f0b298c504daae9da171bac950c036432cc00201
+_commit_armtf=f928897fa5df6afa0f70bb1b1011e4bb03cb4596
```
I was using crust's arm TF fork because it has optional patches for improved compatibility with crust, but I can probably include these patches as part of the package. Either way, I'll try to get upstream uboot with the pine patches. I believe this is what archlinux arm does as well.

jjsullivan5196 replied:
So I tried to use upstream u-boot like I said, but it made 20+ patches so I stuck with the downstream repo. (See review thread before this one)

However, crust's version of armtf only had 5 patches, so the template now uses armtf from ARM Software upstream. Patches from crust are applied at crust's branch point ((commit before d0f999c)[https://github.com/crust-firmware/arm-trusted-firmware/commits/crust]), as it's more recent than ARM's latest release (April 20th).

jjsullivan5196 commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,110 @@
+# please get sources for any new versions from PINE64's u-boot repo and Crust's
+# branch of arm-trusted-firmware.
+
+_commit_uboot=f0b298c504daae9da171bac950c036432cc00201
```
Yes, I'll see if that works.

jjsullivan5196 replied:
u-boot template now uses the `crust-2020-06-12` tag to reference the u-boot sources from pine64's repo

jjsullivan5196 commented on srcpkgs/cross-or1k-none-elf-gcc/template:
```diff
@@ -0,0 +1,91 @@
+nostrip_files="libgcc.a libgcov.a"
+
+_32bit_failed="https://travis-ci.org/github/void-linux/void-packages/jobs/717355887"
+case $XBPS_TARGET_MACHINE in
```
Thanks, will fix this.

jjsullivan5196 replied:
Fixed

sgn commented on srcpkgs/cross-or1k-none-elf-gcc/template:
```diff
@@ -0,0 +1,91 @@
+}
+
+pre_configure() {
+	export CFLAGS_FOR_TARGET="-g -Os -ffunction-sections -fdata-sections"
+	export CXXFLAGS_FOR_TARGET="-g -Os -ffunction-sections -fdata-sections"
```
While you're at it, move this in global scope, remove `pre_configure` and `pre_build`.

jjsullivan5196 replied:
Fixed

jjsullivan5196 commented:
Force-pushed for review changes and new kernel stable version 5.8.3

jjsullivan5196 commented:
Force pushed to enable usb monitor module (enabled in mainline kernel config)

jjsullivan5196 commented:
Forgot to mark bluetooth firmware nonfree, force pushed to fix

jjsullivan5196 commented:
Force push for kernel version 5.8.5

sgn commented:
On 2020-08-28 02:41:25-0700, John Sullivan <notifications@github.com> wrote:
> Forgot to mark bluetooth firmware nonfree, force pushed to fix @sgn 

We usually not send firmware to nonfree (let's see other people's
opinion)
If you think it should go to nonfree, remove it from pinephone-base's
depends too.

-- 
Danh

jjsullivan5196 commented:
That sounds fair, I can revert the change if the opinion is favorable. I'll leave this for context and a few points https://www.pine64.org/2020/01/24/setting-the-record-straight-pinephone-misconceptions/

- This is firmware under a shareware license, so free redistribution is explicitly allowed
- The firmware is uploaded to a wireless NIC via USB, the NIC/WIFI/BT are connected via USB and never access the SoC/RAM directly
- WiFi/bluetooth do not work without the realtek firmware

I made the change in a strict interpretation of free software, but I think these are fairly critical features. I think I'd prefer this be in the main repo, if it's customary to do so.

jjsullivan5196 commented:
@the-maldridge made the argument that the shareware license is pretty restrictive, and this firmware isn't covered by the kernel firmware repo. I think that's enough of an argument to put it in nonfree, so I'm committing to that. `pinephone-base` no longer depends on the realtek firmware. I also made the firmware template more generally applicable for other devices incase it's needed elsewhere.

jjsullivan5196 commented:
Added a udev rule that adds aliases for the modem device nodes in `/dev/`, sourced from https://github.com/dreemurrs-embedded/Pine64-Arch/blob/master/PKGBUILDS/pine64/danctnix-eg25-misc/90-modem-eg25.rules

ericonr commented:
Any chance these rules could be upstreamed? Shipping our own rules shouldn't be necessary for most cases.

jjsullivan5196 commented:
I can't find any repos from pine64 that provide this rule, but the rule itself will work with any devices made from the current supply chain [(available until ~2024)](https://wiki.pine64.org/index.php?title=PinePhone) after then PINE64 will probably be shipping a different product.

I'd get it from upstream if it were available, but I think things like this are inconsequential to PINE64 themselves. I'm fine with adding some docs as a replacement if this isn't advisable.

jjsullivan5196 commented:
pine64's u-boot repo recently added a patch used in this PR, I've changed the source to this commit and dropped the patch https://gitlab.com/pine64-org/u-boot/-/commit/072c0577d8a5ffb0f216a73d430f3f3d800966ba

CameronNemo commented on srcpkgs/rtl8723bt-firmware/template:
```diff
@@ -0,0 +1,27 @@
+version=0.0.202008
+revision=1
+_commit=8840b1052b4ee426f348cb35e4994c5cafc5fbbd
+archs=noarch
```
do not add new packages with noarch defined, please

jjsullivan5196 replied:
gone :+1:

CameronNemo commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,105 @@
+
+_uboot=u-boot-$_tag_uboot
+_armtf=arm-trusted-firmware-$_commit_armtf
+_crust=crust-$_commit_crust
```
We usually try to use released tags rather than direct commits. Seems like it should be easy to do for crust at least.

jjsullivan5196 replied:
It seems the latest stable version of crust is only in master (check [postmarketOS's](https://gitlab.com/postmarketOS/pmaports/-/blob/master/main/crust/APKBUILD) build script) so I can't change to a tag for the moment. However, I've asked crust's developer if a new tag can be added. I'll come back here if I hear anything.

CameronNemo replied:
most recent crust tag seems recent enough from my perspective. ATF is fine as is.

jjsullivan5196 replied:
Cool. I'll just leave this here for posterity: https://github.com/crust-firmware/crust/commit/65fa3f23621aa3a0f4ea48465df629201de954ba This is tag 0.2, with the latest commit being from last year. I'm fairly certain there's been quite a bit done since then.

sgn replied:
Does this mean we expect a new push soon? If yes, we should wait a bit, meanwhile, I have some comment.

jjsullivan5196 replied:
@sgn I tried pinging smaeul (crust dev) on irc but they didn't come around. I can try again.

sgn replied:
Let's wait for a few more days :-p I was about to fix the `-j $makejobs` myself if I didn't see this :-p
Hope you don't mind.

jjsullivan5196 replied:
@sgn Already fixed the makejobs thing, I'll try leaving an issue on their repo too

CameronNemo replied:
Sorry I assumed that the 0.2 tag, which was made on July 12, included all of the commits up to July 11. My bad here.

jjsullivan5196 replied:
No trouble, it should probably be tagged anyway. I was just inclined to go with the flow of what other maintainers were doing. sgn is probably right to hold off to see if anything changes

jjsullivan5196 replied:
@sgn @CameronNemo got a response from [smaeul:](https://view.matrix.org/room/!ToPtLQdvveNEbXcHqE:fam-ribbers.com/$RXeBCeUBvyco7tcUQPlFWtc7qLig2Yq8Xn3syRNycMg?via=matrix.org&via=feneas.org&via=kde.org) `
jjsullivan: crust master is what you should be using, yes. I don't plan to tag 0.3 until I have fixed the known issues with the current version (crashing on a couple of devices)`

So the new tag is probably a bit far off from now, but master is working with the pinephone. The 'other devices' are other Allwinner products using the same embedded controller.

sgn commented:
Ready? I guess

jjsullivan5196 commented:
> Ready? I guess

Yes, the person maintaining the downstream kernel added the 5.8.6 stable bump recently, so I'll commit that now. Everything else should be steady at this point.

jjsullivan5196 commented:
@sgn Ready to go :+1:

sgn commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,105 @@
+	export BL31=$wrksrc/$_armtf/build/sun50i_a64/debug/bl31.bin
+	export SCP=$wrksrc/$_crust/build/scp/scp.bin
+
+	make -C $_uboot $_cross -j $makejobs
```
`$makejobs` has `-j` already.

jjsullivan5196 replied:
did I say ready? My bad :^) All good now

jjsullivan5196 commented:
Marking the crust tag issue as resolved, see comment from the crust author here https://github.com/void-linux/void-packages/pull/24072#discussion_r484476229

julio641742 commented:
Hello, I got my pinephone and tested this today.
I could not boot because 'dracut' is missing and I needed it to generate the initramfs.
Second, I needed to add the 'root=UUID=<id>' in 'pinephone-uboot/files/kernel.d/uboot' '/etc/defaults/pine-iforgot-uboot' because on boot up dracut did not find 'root' even though the generated cmdline did include it.

sgn commented:
> Hello, I got my pinephone and tested this today.
> I could not boot because 'dracut' is missing and I needed it to generate the initramfs.
> Second, I needed to add the 'root=UUID=' in 'pinephone-uboot/files/kernel.d/uboot' '/etc/defaults/pine-iforgot-uboot' because on boot up dracut did not find 'root' even though the generated cmdline did include it.

So, this is not ready?

jjsullivan5196 commented:
@sgn this is an issue with mklive and user configuration, not with this PR in particular. I was going to submit a PR to mklive separately.

Also @julio641742, yes adding `root=PARTUUID=` to `/etc/default/pinephone-uboot-config` is a required step. It can be done by mklive but it needs to be patched in.

jjsullivan5196 commented:
Bump to kernel 5.8.8

sgn commented on srcpkgs/cross-or1k-none-elf/template:
```diff
@@ -0,0 +1,11 @@
+# Template file for 'cross-or1k-none-elf'
+_triplet=or1k-none-elf
+pkgname=cross-or1k-none-elf
+version=0.6
+revision=1
+build_style=meta
+depends="cross-or1k-none-elf-binutils cross-or1k-none-elf-gcc"
+short_desc="GNU cross bare metal toolchain"
+maintainer="John Sullivan <jsullivan@csumb.edu>"
+license="GFDL-1.2-or-later, GPL-3.0-or-later, LGPL-2.1-or-later"
+homepage="https://www.voidlinux.org/"
```
This commit was merged without this package.

sgn commented on srcpkgs/pinephone-uboot/files/kernel.d/uboot:
```diff
@@ -0,0 +1,55 @@
+. /etc/default/pinephone-uboot-config
+
+# Add bootargs for state of volume keys (for extra boot options based on initramfs)
+test $ROCKER_ARGS && ROCKER_TEST=$(cat <<EOF 
```
I'll commit without the space after `EOF`

sgn commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,105 @@
+revision=1
+archs="aarch64*"
+create_wrksrc=yes
+hostmakedepends="flex cross-or1k-none-elf dtc python3 python3-devel bc swig"
```
and `s/cross-or1k-none-elf/&-gcc/`

sgn commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,105 @@
+
+do_patch() {
+	apply_patches() {
+		for _patch in $PATCHESDIR/${1}/*.patch; do
+			patch -d ${2} -N -p1 -i $_patch
+		done
+	}
+
+	apply_patches armtf $_armtf
```
then unroll this function,

sgn commented on srcpkgs/pinephone-uboot/template:
```diff
@@ -0,0 +1,105 @@
+
+	make -C $_crust $_hostcc $_or1k pinephone_defconfig
+	make -C $_uboot $_cross pinephone_defconfig
+	echo 'CONFIG_IDENT_STRING=" Voidlinux"' >> $_uboot/.config
```
with `s/Voidlinux/VoidLinux/`, please complain if anything was wrong from my side.

jjsullivan5196 commented:
@sgn all clear from my end, your help has been greatly appreciated!
